### ISSUE TITLE
Type inference error with generics and function 'this' typing

### ISSUE THREAD

**Issue by use-strict** [Microsoft/TypeScript#9673]
**TypeScript Version:**  typescript@2.1.0-dev.20160713

**Code**

```ts
interface JQuery {
    each<T>(
        collection: T[],
        callback: (this: T, indexInArray: number, valueOfElement: T) => any
    ): any;
}

let $: JQuery;
let list: string[];
$.each(list, function() {
    return this != 'abc';
});

```

**Expected behavior:**
No errors
**Actual behavior:** 
Operator '!=' cannot be applied to types 'T' and 'string'.

Notes:
- Explicit 'string' type parameter fixes the error: `$.each<string>(...)`, although the type is correctly inferred to string even without it.

**Comment by sandersn** [comment]
Also, replacing `return this != 'abc'` with `return this.indexOf('x') != 0` works. It seems related specifically to the binary operator with `this`.

**Comment by sandersn** [comment]
Well, that's only VS Code apparently. Now that I'm running a test case from the command line, `this.indexOf('x')` fails with the same error.

**Comment by sandersn** [comment]
A fix is up at #9746, although I need some expert help to make sure it's the correct fix (it's hacky, so it's probably not correct yet).
